x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/ChainingInputStream.java

```diff
@@ -0,0 +1,214 @@
+    protected InputStream markIn; // protected for tests
+    /**
+     * {@code true} if {@link #close()} has been called; any subsequent {@code read}, {@code skip}
+     * {@code available} and {@code reset} calls will throw {@code IOException}s
+     */
+    private boolean closed;
+
+    /**
+     * This method is responsible for generating the component input streams.
+     * It is passed the current input stream and must return the successive one,
+     * or {@code null} if the current component is the last one. It is passed the {@code null} value
+     * at the very start, when no component input stream has yet been obtained.
+     */
+    abstract @Nullable InputStream nextComponent(@Nullable InputStream currentComponentIn) throws IOException;
+
+    @Override
+    public int read() throws IOException {
+        ensureOpen();
+        do {
+            int byteVal = currentIn == null ? -1 : currentIn.read();
+            if (byteVal != -1) {
+                return byteVal;
+            }
+        } while (nextIn());
+        return -1;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        Objects.checkFromIndexSize(off, len, b.length);
+        if (len == 0) {
+            return 0;
+        }
+        do {
+            int bytesRead = currentIn == null ? -1 : currentIn.read(b, off, len);
+            if (bytesRead != -1) {
+                return bytesRead;
+            }
+        } while (nextIn());
+        return -1;
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        ensureOpen();
+        if (n <= 0) {
+            return 0;
+        }
+        if (currentIn == null) {
+            nextIn();
+        }
+        long bytesRemaining = n;
+        while (bytesRemaining > 0) {
+            long bytesSkipped = currentIn.skip(bytesRemaining);
+            if (bytesSkipped == 0) {
+                int byteRead = read();
+                if (byteRead == -1) {
+                    break;
+                } else {
+                    bytesRemaining--;
+                }
+            } else {
+                bytesRemaining -= bytesSkipped;
+            }
+        }
+        return n - bytesRemaining;
+    }
+
+    @Override
+    public int available() throws IOException {
+        ensureOpen();
+        if (currentIn == null) {
+            nextIn();
+        }
+        return currentIn.available();
+    }
+
+    @Override
+    public boolean markSupported() {
+        return true;
+    }
+
+    @Override
+    public void mark(int readlimit) {
+        if (markSupported() && false == closed) {
+            // closes any previously stored mark input stream
+            if (markIn != null && markIn != EXHAUSTED_MARKER && currentIn != markIn) {
+                try {
+                    markIn.close();
+                } catch (IOException e) {
+                    // an IOException on a component input stream close is not important
```

Comment:
I have opted for logging the exception. Given the ephemeral nature of the component streams (that a **new** instance is returned for each invocation of `#nextComponent` and that the same segment of the overall byte stream can be observed through several instances - the mark/reset case) I don't see the IOException during a close as significant. In addition the caller wouldn't have any other option of handling the exception but to abort any further operations on the chaining input stream, but this is unnecessary IMO because such closed components are never reused and they go out of scope immediately.

If you believe only logging could be hiding bugs, we can throw `UncheckedIOException` or least preferred, I can make the `IOException` be thrown during other method calls, but that would complicate the code a little.

Reply:
I other words, I expect `close` is a no-op for components.